### PR TITLE
Add a cost to all conditions and a new "free conditions" counter

### PR DESCRIPTION
**Pull request by matt-o-how**
This PR adds a new flag called COST_CONDITIONS which is activated after hard fork 2 which adds a cost to all conditions after the hard fork. There are a number of "free conditions" for every spend, which is stored as constant in opcodes set to 100.

### CODE REVIEW

**Review by arvidn — CHANGES_REQUESTED** [review]
you need tests for this as well. positive tests, edge cases and negative tests

File: crates/chia-consensus/src/gen/opcodes.rs [hunk 1]
```diff
@@ -64,6 +64,10 @@ pub const SOFTFORK: ConditionOpcode = 90;
 
 pub const CREATE_COIN_COST: Cost = 1_800_000;
 pub const AGG_SIG_COST: Cost = 1_200_000;
+
+pub const GENERIC_CONDITION_COST: Cost = 1_200_000;
```
Comment (arvidn) [inline crates/chia-consensus/src/gen/opcodes.rs:68]: I think this should be in the range 100-500.

File: crates/chia-consensus/src/gen/test_generators.rs [hunk 2]
```diff
@@ -2,7 +2,7 @@ use super::conditions::{NewCoin, SpendBundleConditions, SpendConditions};
 use super::run_block_generator::{run_block_generator, run_block_generator2};
 use crate::allocator::make_allocator;
 use crate::consensus_constants::TEST_CONSTANTS;
-use crate::gen::flags::{DONT_VALIDATE_SIGNATURE, MEMPOOL_MODE};
+use crate::r#gen::flags::{DONT_VALIDATE_SIGNATURE, MEMPOOL_MODE};
```
Comment (arvidn) [inline crates/chia-consensus/src/gen/test_generators.rs:5]: what does this do?

File: crates/chia-consensus/src/spendbundle_validation.rs [hunk 3]
```diff
@@ -2,6 +2,7 @@ use crate::allocator::make_allocator;
 use crate::consensus_constants::ConsensusConstants;
 use crate::gen::owned_conditions::OwnedSpendBundleConditions;
 use crate::gen::validation_error::ErrorCode;
+use crate::r#gen::flags::COST_CONDITIONS;
```
Comment (arvidn) [inline crates/chia-consensus/src/spendbundle_validation.rs:5]: if there's some benefit to `r#`, I imagine we should add it to every `use`, but probably not in this PR

File: crates/chia-consensus/src/gen/conditions.rs [hunk 4]
```diff
@@ -4668,50 +4744,104 @@ fn test_softfork_condition_failures(#[case] conditions: &str, #[case] expected_e
...
+    ASSERT_CONCURRENT_PUZZLE,
+    1025,
+    COST_CONDITIONS,
     Some(ErrorCode::AssertConcurrentPuzzleFailed)
 )]
-#[case(ASSERT_CONCURRENT_PUZZLE, 1025, Some(ErrorCode::TooManyAnnouncements))]
 fn test_limit_announcements(
     #[case] cond: ConditionOpcode,
     #[case] count: i32,
+    #[case] flag: u32,
```
Comment (arvidn) [inline crates/chia-consensus/src/gen/conditions.rs:4839]: in `rstest` you can also add an independent parameter by saying:
```
#[values(0, COST_CONDITIONS)]
```
Then every test is run in both configuration. You could check the result programatically.

File: crates/chia-consensus/src/gen/conditions.rs [hunk 5]
```diff
@@ -1211,27 +1227,39 @@ pub fn parse_conditions<V: SpendVisitor>(
...
             Condition::CreateCoinAnnouncement(msg) => {
-                decrement(&mut announce_countdown, msg)?;
+                if (flags & COST_CONDITIONS) == 0 {
+                    decrement(&mut announce_countdown, msg)?;
+                };
                 state.announce_coin.insert((spend.coin_id.clone(), msg));
             }
             Condition::CreatePuzzleAnnouncement(msg) => {
-                decrement(&mut announce_countdown, msg)?;
+                if flags & COST_CONDITIONS != COST_CONDITIONS {
```
Comment (arvidn) [inline crates/chia-consensus/src/gen/conditions.rs:1236]:
```suggestion
                if flags & COST_CONDITIONS == 0 {
```
Perhaps this is a matter of taste. I prefer this one, and right now it's not consistent, which looks a bit odd.

File: crates/chia-consensus/src/gen/conditions.rs [hunk 6]
```diff
@@ -1039,7 +1043,19 @@ pub fn parse_conditions<V: SpendVisitor>(
                 *max_cost -= AGG_SIG_COST;
                 ret.condition_cost += AGG_SIG_COST;
             }
-            _ => (),
+            _ => {
+                if (flags & COST_CONDITIONS) != 0 {
```
Comment (arvidn) [inline crates/chia-consensus/src/gen/conditions.rs:1047]: maybe it was more reasonable the way you had it before, where it always applied to all conditions. I suppose we'll have to update a lot of tests that way. But the `SOFTFORK` condition causes complications with this

File: crates/chia-consensus/src/gen/opcodes.rs (hunk 1 above)
Comment (matt-o-how) [reply crates/chia-consensus/src/gen/opcodes.rs]: Set to 500

File: crates/chia-consensus/src/gen/conditions.rs [hunk 7]
```diff
@@ -1493,7 +1524,7 @@ pub fn validate_conditions(
     }
 
     if !state.assert_concurrent_puzzle.is_empty() {
-        let mut spent_phs = HashSet::<Bytes32>::with_capacity(state.spent_puzzles.len());
```
[the remark this answers is not in the document]
Comment (matt-o-how) [reply crates/chia-consensus/src/gen/conditions.rs]: reverted

File: crates/chia-consensus/src/gen/conditions.rs (hunk 5 above)
Comment (matt-o-how) [reply crates/chia-consensus/src/gen/conditions.rs]: standardised

File: crates/chia-consensus/src/gen/conditions.rs (hunk 6 above)
Comment (matt-o-how) [reply crates/chia-consensus/src/gen/conditions.rs]: implemented

File: crates/chia-consensus/src/spendbundle_validation.rs (hunk 3 above)
Comment (arvidn) [reply crates/chia-consensus/src/spendbundle_validation.rs:5]: I think you should remove these

File: crates/chia-consensus/src/gen/conditions.rs [hunk 8]
```diff
@@ -4668,50 +4844,104 @@ fn test_softfork_condition_failures(#[case] conditions: &str, #[case] expected_e
...
+    ASSERT_CONCURRENT_PUZZLE,
+    1025,
+    COST_CONDITIONS,
     Some(ErrorCode::AssertConcurrentPuzzleFailed)
 )]
-#[case(ASSERT_CONCURRENT_PUZZLE, 1025, Some(ErrorCode::TooManyAnnouncements))]
 fn test_limit_announcements(
     #[case] cond: ConditionOpcode,
     #[case] count: i32,
+    #[case] flag: u32,
```
Comment (arvidn) [inline crates/chia-consensus/src/gen/conditions.rs:4939]:
```suggestion
    #[values(0, COST_CONDITIONS)] flag: u32,
```

**Review by arvidn — APPROVED** [review]
